rotkehlchen/tasks/historical_balances.py

```diff
@@ -0,0 +1,242 @@
+            """,
+            (EventMetricKey.BALANCE, from_ts),
+        )
+        for row in cursor:
+            if (bucket := Bucket.from_db(row[:4])) not in bucket_balances:
+                bucket_balances[bucket] = FVal(row[4])
+
+    log.debug(f'Loaded {len(bucket_balances)} bucket balances before ts={from_ts}')
+    return bucket_balances
+
+
+def process_historical_balances(
+        database: 'DBHandler',
+        msg_aggregator: 'MessagesAggregator',
+        from_ts: Timestamp | None = None,
+) -> None:
+    """Process events and compute balance metrics. Stops on negative balance."""
+    log.debug(f'Starting historical balance processing from_ts={from_ts}')
+    bucket_balances = {}
+    if from_ts is not None:
+        bucket_balances = _load_bucket_balances_before_ts(database, from_ts)
+
+    with database.conn.read_ctx() as cursor:
+        filter_query = HistoryEventFilterQuery.make(
+            from_ts=from_ts,
+            order_by_rules=[('timestamp', True), ('sequence_index', True)],
+            exclude_ignored_assets=True,
+        )
+        events = DBHistoryEvents(database).get_history_events_internal(
+            cursor=cursor,
+            filter_query=filter_query,
+        )
+
+    if (total_events := len(events)) == 0:
+        log.debug('No events to process for historical balances')
+        _finalize_processing(database)
+        return
+
+    metrics_batch: list[tuple[int | None, str | None, str, str]] = []
+    first_batch_written, negative_balance_buckets = False, set[Bucket]()
+    send_ws_every = msg_aggregator.how_many_events_per_ws(total_events)
+    for idx, event in enumerate(events):
+        if (bucket := Bucket.from_event(event)) in negative_balance_buckets:
+            continue
+
+        current_balance_in_bucket = bucket_balances.get(bucket, ZERO)
+        if (direction := event.maybe_get_direction()) == EventDirection.IN:
+            new_balance = current_balance_in_bucket + event.amount
+        elif direction == EventDirection.OUT:
+            if (new_balance := current_balance_in_bucket - event.amount) < ZERO:
+                msg_aggregator.add_message(
+                    message_type=WSMessageType.NEGATIVE_BALANCE_DETECTED,
+                    data={
+                        'event_identifier': event.identifier,
+                        'group_identifier': event.group_identifier,
+                        'asset': event.asset.identifier,
+                        'balance_before': str(current_balance_in_bucket),
+                    },
+                )
+                log.warning(
+                    f'Negative balance detected for {event.asset.identifier} '
+                    f'at event {event.identifier}. Skipping {bucket}.',
+                )
+                negative_balance_buckets.add(bucket)
+                continue
+        else:
+            new_balance = current_balance_in_bucket
+
+        bucket_balances[bucket] = new_balance
```

Comment:
I would say to keep this in the two branches that you need it. Otherwise you are assigning when it is not needed for neutral events